pkg/relayer/proxy/synchronous.go

```diff
@@ -244,27 +245,64 @@ func (sync *synchronousRPCServer) serveHTTP(
 		return nil, err
 	}
 
-	// Get the relayRequest payload's `io.ReadCloser` to add it to the http.Request
-	// that will be sent to the proxied (i.e. staked for) service.
-	// (see https://pkg.go.dev/net/http#Request) Body field type.
-	requestBodyReader := io.NopCloser(bytes.NewBuffer(relayRequest.Payload))
-	defer requestBodyReader.Close()
-	sync.logger.Debug().
-		Str("request_payload", string(relayRequest.Payload)).
-		Msg("serving relay request")
+	// Deserialize the relay request payload to get the upstream HTTP request.
+	upstreamRequest, err := httpcodec.DeserializeHTTPRequest(relayRequest.Payload)
+	if err != nil {
+		return nil, err
+	}
 
 	// Build the request to be sent to the native service by substituting
 	// the destination URL's host with the native service's listen address.
 	sync.logger.Debug().
 		Str("destination_url", serviceConfig.BackendUrl.String()).
 		Msg("building relay request payload to service")
 
+	// Parse the upstream request URL then replace the host and scheme with the
+	// service's backend URL's.
+	requestUrl, err := url.Parse(upstreamRequest.URL)
+	if err != nil {
+		return nil, err
+	}
+	requestUrl.Host = serviceConfig.BackendUrl.Host
+	requestUrl.Scheme = serviceConfig.BackendUrl.Scheme
+
+	// Prepend the path of the service's backend URL to the path of the upstream request.
```

Comment:
My repeating question is what should be in the shannon SDK vs appgateserver.

For example, will other gateways need to copy-pasta or be aware of this logic? If so, should it be in the shannon SDK?

Reply:
This is a `RelayMiner` specific logic, `SDK` and `AppGateServer`s do not need to be aware of it.

It is basically the logic to translate an `AppGateServer`->`RelayMiner` request to a `RelayMiner`->`BackendService` request.

The `ApggateServer` doesn't have any knowledge about the routing internals of the `RelayMiner` but need to receive a reply "as if" it sent the request directly to the `BackendService`.

Beside the request payload, the `RelayMiner` need to forward the headers and path to the `BackendService`.

Reply:
Great context. #PUC